text/0000-list-rendering-reform.md

```diff
@@ -0,0 +1,306 @@
+---
+title: List rendering reform
+status: DRAFTED
+created_at: 2023-11-14
+updated_at: 2023-11-14
+pr: https://github.com/salesforce/lwc-rfcs/pull/84
+---
+
+# List rendering reform
+
+## Summary
+
+This RFC proposes a new template directive, `lwc:each`, which resolves issues and provides enhancements compared to the existing `for:each` and `iterator:*` directives.
+
+## Basic example
+
+```html
+<template>
+  <ul>
+    <template lwc:each={contacts} 
+              lwc:item="contact"
+              lwc:key={contact.id}
+              lwc:index="index"
+              lwc:first="first"
+              lwc:last="last"
+              lwc:even="even"
+              lwc:odd="odd">
+      <li>
+        Name: {contact.name}
+        Index: {index}
+        First? {first}
+        Last? {last}
+        Even? {even}
+        Odd? {odd}
+      </li>
+    </template>
+  </ul>
+</template>
+```
+
+## Motivation
+
+This proposal supersedes the existing [`for:each`](https://lwc.dev/guide/html_templates#for%3Aeach) and [`iterator:*`](https://lwc.dev/guide/html_templates#iterator) directives and resolves several issues with them:
+
+- We have two directives for lists instead of one, with different features available on each. The new directive unifies them.
+- The existing directives do not use the `lwc:` prefix, unlike every other modern LWC directive. Using `lwc:` makes it clear which attributes are LWC-specific, and would allow for improvements in the future such as [shorthands](https://github.com/salesforce/lwc/issues/3303) (out of scope for this RFC).
+- The key should be defined on the list root element, not on each element inside the list. The current behavior requires tediously repeating the `key` for each element inside the list, and also leads to [inconsistent behavior](https://github.com/salesforce/lwc/issues/3860) for text/comment nodes.
+- There is no way to easily render based on even vs odd list items.

```

Comment:
@pmdartus raised another benefit: we could use Fragments as the root of each list item. This should make the vdom diffing much more efficient because we would be diffing fewer vnodes in cases where the list repeater has multiple top-level elements.

Today we can't do this due to `key`. A developer could technically put a different `key` on every top-level element inside of a list, which would defeat the fragment optimization since the fragment must have one key. The new proposal solves this since `lwc:key` is hoisted.